tests/flattenTest.php

```diff
@@ -0,0 +1,189 @@
+					[2, 3],
+					[7,9],
+					8,
+				],
+				'expected' => [4, 2, 3, 7, 9, 8]
+			],
+			'With an indexed array with only array values' => [
+				'iterable' => [
+					[2, 3],
+					[7,9],
+				],
+				'expected' => [2, 3, 7, 9]
+			],
+
+			/*
+				With associative array
+			 */
+
+			'With an associative array with only scalar values' => [
+				'iterable' => [
+					'a' => 6,
+					'b' => 1,
+					'c' => 3,
+				],
+				'expected' => [ 6, 1, 3]
+			],
+			'With an associative array with a mix of scalar and array values' => [
+				'iterable' => [
+					'a' => 4,
+					'b' => [2, 3],
+					'c' => [7,9],
+					'd' => 8,
+				],
+				'expected' => [4, 2, 3, 7, 9, 8]
+			],
+			'With an associative array with only array values' => [
+				'iterable' => [
+					'b' => [2, 3],
+					'c' => [7,9],
+				],
+				'expected' => [2, 3, 7, 9]
+			],
+
+			/*
+				With stdClass
+			 */
+
+			'With an stdClass with only scalar values' => [
+				'iterable' => (object) [
+					'a' => 6,
+					'b' => 1,
+					'c' => 3,
+				],
+				'expected' => [6, 1, 3]
+			],
+			'With an stdClass with a mix of scalar and array values' => [
+				'iterable' => (object) [
+					'a' => 4,
+					'b' => [2, 3],
+					'c' => [7,9],
+					'd' => 8,
+				],
+				'expected' => [4, 2, 3, 7, 9, 8]
+			],
+			'With an stdClass with only array values' => [
+				'iterable' => (object) [
+					'b' => [2, 3],
+					'c' => [7,9],
+				],
+				'expected' => [2, 3, 7, 9]
+			],
+
+			/*
+				With ArrayObject
+			 */
+
+			'With an ArrayObject with only scalar values' => [
+				'iterable' => new ArrayObject([
+					'a' => 6,
+					'b' => 1,
+					'c' => 3,
+				]),
+				'expected' => [6, 1, 3]
+			],
+			'With an ArrayObject with a mix of scalar and array values' => [
+				'iterable' => new ArrayObject([
+					'a' => 4,
+					'b' => [2, 3],
+					'c' => [7,9],
+					'd' => 8,
+				]),
+				'expected' => [4, 2, 3, 7, 9, 8]
+			],
+			'With an ArrayObject with only array values' => [
+				'iterable' => new ArrayObject([
+					'b' => [2, 3],
+					'c' => [7,9],
+				]),
+				'expected' => [2, 3, 7, 9]
+			]
+		];
+	}
+
+	/**
+	 * @dataProvider casesTypeAssertions
+	 * @expectedException InvalidArgumentException
+	 */
+	public function testTypeAssertions($iterable, $type)
+	{
+		try {
+			Dash\flatten($iterable);
+		}
+		catch (Exception $e) {
+			$this->assertSame(
+				"Dash\\flatten expects iterable or stdClass or null but was given $type",
+				$e->getMessage()
+			);
+			throw $e;
+		}
+	}
+
+	public function casesTypeAssertions()
+	{
+		return [
+			'With an empty string' => [
+				'iterable' => '',
+				'type' => 'string',
+			],
+			'With a string' => [
+				'iterable' => 'hello',
+				'type' => 'string',
+			],
+			'With a zero number' => [
+				'iterable' => 0,
+				'type' => 'integer',
+			],
+			'With a number' => [
+				'iterable' => 3.14,
+				'type' => 'double',
+			],
+			'With a DateTime' => [
+				'iterable' => new DateTime(),
+				'type' => 'DateTime',
+			],
+		];
+	}
+}
```

Comment:
Missing: test case that tests the examples from the php docblock. It might seem silly, but I’ve encountered enough cases where doing so catches gaps in logic. It’s also a loose guarantee that the documentation is still accurate.

Reply:
Done